keps/sig-api-machinery/3937-declarative-validation/README.md

```diff
@@ -0,0 +1,1077 @@
+  enabled, any difference in the validation errors caught by the two validators
+  will be logged in a special way making them easy to find and analyze, but only
+  the validation errors caught by legacy validation will be reported to the
+  client.
+- e2e tests will be instrumented to detect any difference in validation rules
+  between declarative validation and legacy validation, using the log
+  information mentioned above, and to fail if a PR introduces a difference.
+- `validation_test.go` unit tests will be improved to make it easy to verify
+  that declarative validation rules are agreeing with legacy validation rules.
+- A linter will be added to ensure that all versions of an API have the same
+  validation rules unless an exception is made (via a line in an exceptions
+  file). In practice we expect the number of exceptions to be very small and
+  so this linter will primarily ensure that we keep API validation in sync.
+- We will benchmark all validation changes. We will make sure there is a
+  framework in place for this so that it is easy to do.
+
+Migration steps for each API group:
+
+- In a single PR:
+  - Add declarative validation rules using Go struct tags.
+  - Validate that both declarative validation and legacy validation produce
+    the same errors for validation_test.go.
+  - Enable e2e testing testing of the API group so that each e2e test run includes
+    the API group in `DeclarativeValidationEnabled` and checks for any differences
+    reported in logs between declarative validation and legacy validation and
+    fails the e2e test run if any are found.
+  - API reviewers approve the change.
+- ...Time passes. All API validation changes are deamed stable...
+- In a single PR:
+  - The `DeclarativeValidationEnabled` is defaulted to true.
+  - The `LegacyValidationEnabled` is defaulted to false.
+- ...Deprecation of legacy validation is announced...
+- ...Deprecation wait period passes...
+- The legacy API is deleted and can no longer be enabled/disabled via the
+  feature gates.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- Risk: Declarative validation adds significant latency to API request handling.
+  - Potential causes:
+    - Validation versioned types introduces extra conversions.
+      - Why this might be OK: Requests are received as the versioned type, so it
+        should be feasible to avoid extra conversions.
```

Comment:
Would also be a minor security win - currently conversion to internal involves use of `unsafe` operations for performance, and doing validation before conversion reduces the risk of malformed data making it to conversion.